src/js/tests/newsletter-subscription.test.js

```diff
@@ -0,0 +1,95 @@
+
+Object.defineProperty(global.window, 'alert', function fakeAlert(val) {
+  // eslint-disable-next-line no-console
+  console.log(val);
+});
+
+const NewsletterSubscription = require('../newsletter-subscription');
+
+const exampleFormHTML = `
+<form id="newsletter-form" class="mzp-c-newsletter-form" name="newsletter-form">
+  <div class="js-newsletter-fields" data-success-message="<b>Thank you. Please check your email to confirm your subscription.</b>">
+    <input type="hidden" name="newsletters" value="app-dev">
+    <input type="hidden" name="fmt" value="H">
+    <p>
+      <label for="newsletter-email">Your email address</label>
+      <input type="email" name="email" class="newsletter-email" id="newsletter-email" placeholder="you@example.com" required aria-required="true">
+    </p>
+    <p>
+      <label for="newsletter-privacy" class="mzp-u-inline">
+        <input type="checkbox" name="privacy" id="newsletter-privacy" required aria-required="true">
+        I'm okay with Mozilla handling my info as explained in this <a class="privacy-link" href="https://www.mozilla.org/privacy/websites/" target="_blank" rel="noopener noreferrer">Privacy Notice</a>.
+      </label>
+    </p>
+    <p>
+      <button type="submit" class="mzp-c-button mzp-t-small">Sign up now</button>
+    </p>
+  </div>
+</form>`;
+
+/** Mock out the Fetch and return custom responses for 'https://www.mozilla.org/en-US/newsletter/';
+ */
+beforeEach(() => {
+  fetchMock.mockIf(/^https?:\/\/mozilla.org.*$/, req => {
+    if (req.url.endsWith('/en-US/newsletter/')) {
+      return { body: { status: 'OK' } };
+    }
+    if (req.url.endsWith('/en-US/something-else')) {
+      return {
+        body: { status: 'WRONG URL' },
+      };
+    }
+    return {
+      status: 404,
+      body: 'Not Found',
+    };
+  });
+});
+
+/**
+ * Helper to boostrap form with test data
+ *
+ *  @param {document} document
+ *  @param {object} data
+ */
+function populateFormData(document, data) {
+  const emailField = document.getElementsByName('email')[0];
+  const privacyField = document.getElementsByName('privacy')[0];
+  emailField.value = data.email;
+  privacyField.value = data.privacy;
+}
+
+test('Happy path', () => {
+  document.body.innerHTML = exampleFormHTML;
+  NewsletterSubscription.init(); // this is how it's invoked in index.js
+  const form = document.getElementById('newsletter-form');
+
+  populateFormData(document, {
+    email: 'test@example.com',
+    privacy: 'checked',
+  });
+  form.submit();
+
+  expect(fetch.mock.calls[0][1].headers).toEqual({
+    'X-Requested-With': 'XMLHttpRequest',
+    'Content-type': 'application/x-www-form-urlencoded',
+  });
+  expect(fetch.mock.calls[0][1].method).toEqual('POST');
+  expect(fetch.mock.calls[0][1].body).toEqual(
+    'newsletters=app-dev&fmt=H&email=test%40example.com',
+  );
+  expect(fetch.mock.calls[0][0]).toEqual(
+    'https://www.mozilla.org/en-US/newsletter/',
+  );
+
+  expect(fetch.mock.calls.length).toEqual(2);
```

Comment:
@Gregoor This value should really be `1` in my opinion, unless `fetch` does some double-posting as normal...?

If you run the project locally and look at the network panel when you sign up, you'll see the browser makes two requests, too, so it's not only in Jest that it works (and note that in Jest there is only one DOM element that matches the selector used for binding, etc). So, suggestions about what I've got wrong here would be welcome.

Once I'm clearer about why it's 2 / how to make it work as expected with just 1 POST, I can add some unhappy paths to the tests too.

Reply:
![Screenshot 2020-03-13 at 11 59 00](https://user-images.githubusercontent.com/101457/76619129-12fbb280-6522-11ea-9233-979535f7fece.png)